### PR TITLE
feat: Add gauge metric

### PR DESCRIPTION
**Pull request by mkmik**
This PR implements a Gauge metric + all the testing helpers that will allow us to test the code that uses gauges.

It's correct to think of this gauge implementation as a hack.
It's possible to spend more time and figure out how to do this "the right" way,
but I really think we need gauges right now.

On the upside, I think that the API is going to be stable even once we figure out the "right way" to plumb this into opentelemetry.

### CODE REVIEW

File: metrics/src/lib.rs [hunk 1]
```diff
@@ -268,6 +270,72 @@ impl Domain {
...
+                match unit {
+                    Some(unit) => {
+                        format!("{}.{}", self.build_metric_prefix(name, None), unit)
+                    }
+                    None => self.build_metric_prefix(name, None),
+                },
+                move |arg| {
+                    for i in values_captured.iter() {
+                        // currently rust type inference cannot deduct the type of i.value()
+                        let i: RefMulti<'_, _, (u64, Vec<KeyValue>), _> = i;
```
Comment (e-dard) [inline metrics/src/lib.rs:328]: 🤯

File: metrics/src/lib.rs [hunk 2]
```diff
@@ -581,4 +649,50 @@ mod test {
...
+            &[KeyValue::new("tier", "c"), KeyValue::new("tier", "b")],
+        );
+
+        reg.has_metric_family("http_mem_bytes")
+            .with_labels(&[("tier", "a")])
+            .gauge()
+            .eq(40.0)
+            .unwrap();
+
+        reg.has_metric_family("http_mem_bytes")
```
Comment (e-dard) [inline metrics/src/lib.rs:679]: Could you add one more sample using labels other than `tier` just to show that the default and other labels get merged correctly?

**Review by e-dard — APPROVED** [review]
One suggestion about an extra test but this looks great! Thank you 🙏

File: metrics/src/lib.rs (hunk 2 above)
Comment (mkmik) [reply metrics/src/lib.rs:680]: good idea; done!

**Comment by mkmik** [comment]
merging because I think the API is stable and we can start using it;

comments/improvements to internals are welcome at any time!

**Pull request merged**